common/conf/conf.cpp

```diff
@@ -32,6 +59,71 @@ string GetEnv(const string &var_name, const string &default_value) {
 	}
 }
 
+ExpectedOptionValue CmdlineOptionsIterator::Next() {
+	string option = "";
+	string value = "";
+
+	if (pos_ >= args_.size()) {
+		return ExpectedOptionValue({"", ""});
+	}
+
+	if (past_double_dash_) {
+		OptionValue opt_val {"", args_[pos_]};
+		pos_++;
+		return ExpectedOptionValue(opt_val);
+	}
+
+	if (args_[pos_] == "--") {
+		past_double_dash_ = true;
+		pos_++;
+		return ExpectedOptionValue({"--", ""});
+	}
+
+	if (args_[pos_][0] == '-') {
+		auto eq_idx = args_[pos_].find('=');
+		if (eq_idx != string::npos) {
+			option = args_[pos_].substr(0, eq_idx);
+			value = args_[pos_].substr(eq_idx + 1, args_[pos_].size() - eq_idx - 1);
+			pos_++;
+		} else {
+			option = args_[pos_];
+			pos_++;
+		}
+
+		if (opts_with_value_.count(option) != 0) {
+			// option with value
+			if ((value == "") && ((pos_ >= args_.size()) || (args_[pos_][0] == '-'))) {
+				// the next item is not a value
+				error::Error err = MakeError(
+					ConfigErrorCode::InvalidOptionsError, "Option " + option + " missing value");
+				return ExpectedOptionValue(err);
+			} else if (value == "") {
+				// only assign the next item as value if there was no value
+				// specified as '--opt=value' (parsed above)
+				value = args_[pos_];
+				pos_++;
+			}
+		} else if (opts_wo_value_.count(option) == 0) {
+			// unknown option
+			error::Error err = MakeError(
+				ConfigErrorCode::InvalidOptionsError, "Unrecognized option '" + option + "'");
+			return ExpectedOptionValue(err);
+		} else if (value != "") {
+			// option without a value, yet, there was a value specified as '--opt=value' (parsed
+			// above)
+			error::Error err = MakeError(
+				ConfigErrorCode::InvalidOptionsError,
+				"Option " + option + " doesn't expect a value");
+			return ExpectedOptionValue(err);
+		}
+	} else {
+		value = args_[pos_];
+		pos_++;
+	}
+
+	return ExpectedOptionValue({std::move(option), std::move(value)});
```

Comment:
Q: I think the `move` is unnecessary here, as the values go out of scope, and will be automatically moved.

Reply:
I have to admit I find it a little harder to read when errors and regular expected returns look exactly the same.

Reply:
I suppose this will soon be changed to `unexpected` though, so they won't the same much longer.